microceph/ceph/replication.go

```diff
@@ -0,0 +1,127 @@
+package ceph
+
+import (
+	"context"
+	"fmt"
+	"reflect"
+
+	"github.com/canonical/lxd/shared/logger"
+	"github.com/canonical/microceph/microceph/api/types"
+	"github.com/canonical/microceph/microceph/constants"
+	"github.com/canonical/microceph/microceph/interfaces"
+	"github.com/qmuntal/stateless"
+)
+
+type repArgIndex int
+type ReplicationState string
+
+const (
+	StateDisabledReplication ReplicationState = "replication_disabled"
+	StateEnabledReplication  ReplicationState = "replication_enabled"
+)
+
+const (
+	repArgHandler  repArgIndex = 0
+	repArgResponse repArgIndex = 1
+	repArgState    repArgIndex = 2
+)
+
+type ReplicationHandlerInterface interface {
+	PreFill(ctx context.Context, request types.ReplicationRequest) error
+	GetResourceState() ReplicationState
+	EnableHandler(ctx context.Context, args ...any) error
+	DisableHandler(ctx context.Context, args ...any) error
+	ConfigureHandler(ctx context.Context, args ...any) error
+	ListHandler(ctx context.Context, args ...any) error
+	StatusHandler(ctx context.Context, args ...any) error
+}
+
+func GetReplicationHandler(name string) ReplicationHandlerInterface {
+	// Add RGW and CephFs Replication handlers here.
+	table := map[string]ReplicationHandlerInterface{
+		"rbd": &RbdReplicationHandler{},
+	}
+
+	rh, ok := table[name]
+	if !ok {
+		return nil
+	}
+
+	return rh
+}
+
+func GetReplicationStateMachine(initialState ReplicationState) *stateless.StateMachine {
+	newFsm := stateless.NewStateMachine(initialState)
+	// Configure transitions for disabled state.
+	newFsm.Configure(StateDisabledReplication).
+		Permit(constants.EventEnableReplication, StateEnabledReplication).
+		OnEntryFrom(constants.EventDisableReplication, disableHandler).
+		InternalTransition(constants.EventListReplication, listHandler).
+		InternalTransition(constants.EventDisableReplication, disableHandler)
```

Comment:
Iiuc getting a status from StateDisabled is not supported, is this correct?

Reply:
Yes

Reply:
Ack, guess this edge should be removed then